src/ufuncs.c

```diff
@@ -448,30 +539,119 @@ static void *union_data[1] = {GEOSUnion_r};
 static void *shared_paths_data[1] = {GEOSSharedPaths_r};
 typedef void *FuncGEOS_YY_Y(void *context, void *a, void *b);
 static char YY_Y_dtypes[3] = {NPY_OBJECT, NPY_OBJECT, NPY_OBJECT};
+
+/* There are two inner loop functions for the YY_Y. This is because this
+ * pattern allows for ufunc.reduce.
+ * A reduce operation requires special attention, because we output the
+ * result in a temporary array. NumPy expects that the output array is
+ * filled during the inner loop, so that it can take the first input of
+ * the reduction operation to be the output array. Easiest to show by
+ * example:
+
+ * function called:         out = intersection.reduce(in)
+ * initialization by numpy: out[0] = in[0]; in1 = out; in2[:] = in[:]
+ * first loop:              out[0] = func(out[0], in[1])
+ * second loop:             out[0] = func(out[0], in[2])
+ */
+static void YY_Y_func_reduce(char **args, npy_intp *dimensions,
+                             npy_intp *steps, void *data)
+{
+    FuncGEOS_YY_Y *func = (FuncGEOS_YY_Y *)data;
+    GEOSGeometry *in1, *in2, *out;
+
+    GEOS_INIT_THREADS;
+
+    if (!get_geom(*(GeometryObject **)args[0], &out)) {
+        errstate = PGERR_NOT_A_GEOMETRY;
+    } else if (out != NULL) {
+        BINARY_LOOP {
+            // This is the main reduce logic: in1 becomes previous out
+            if (i > 1) {
+                // If i == 0, in1 is undefined
+                // If i == 1, in1 is the first input, which is owned by python
+                GEOSGeom_destroy_r(ctx, in1);
+            }
+            in1 = out;
+            // Get the other geometry (as normal)
+            if (!get_geom(*(GeometryObject **)ip2, &in2)) {
+                errstate = PGERR_NOT_A_GEOMETRY;
+                GEOSGeom_destroy_r(ctx, out);
+                break;
+            }
+            if (in2 == NULL) {
+                // We break the loop as the answer will remain NULL (None)
+                GEOSGeom_destroy_r(ctx, out);
+                out = NULL;
+                break;
+            } else {
+                out = func(ctx, in1, in2);
+                if (out == NULL) {
+                    errstate = PGERR_GEOS_EXCEPTION;
+                    GEOSGeom_destroy_r(ctx, out);
+                    break;
+                }
+            }
+        }
+    }
+
+    GEOS_FINISH_THREADS;
+
+    // fill the numpy array with a single PyObject while holding the GIL
+    if (errstate == PGERR_SUCCESS) {
+        geom_arr_to_npy(&out, args[2], steps[2], 1);
+    }
+}
+
 static void YY_Y_func(char **args, npy_intp *dimensions,
                       npy_intp *steps, void *data)
 {
+    // A reduce is characterized by the step of the output array being 0:
+    if ((steps[2] == 0) && (args[0] == args[2])) {
```

Comment:
Suggestion: convert this to nested conditionals, which more clearly associates the comment with the primary condition:

```suggestion
    if (steps[2] == 0) {
        if (args[0] == args[2]) {...}
        else if (dimensions[0] > 1) {...}
    }
```